### PR TITLE
RFC: API Change: usb: Make users call usb_enable. Provide global status callback.

### LINKED ISSUE #21419: RFC: API Change: usb: Make users call usb_enable. Provide global status callback.
## Introduction

* Users should have a possibility to call usb_enable when they wants to (like its done for bluetooth). Currently usb is auto enabled on system boot up which may lead to errors (explained later).

* USB status codes should refer to USB device as a whole not to a particular class as it is done now (HID). 

### Problem description

* It is possible that usb is enabled before the class specific init function is called. If that happen usb hardware may start to produce events that will be dropped because class is not initialized yet.

* Until now users were able to view the internal USB driver status codes only by using HID class. In fact usb status is global per a device not per function/class.

### Proposed change

Make users responsible for calling usb_enable and decide whenever hardware should start to work.

Provide global status callback which will indicate the status of USB.

## Detailed 

To solve above problems I propose:
Delete usb_enable() from [usb_device_init()](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/usb_device.c#L1636) which is called on system boot up.

Add a callback parameter to usb_enable(). If user wants to monitor the status of USB he could use this callback.

### Proposed change (Detailed)

Please look to #20375 PR.

### Dependencies

Users that use USB will have to update their applications and call usb_enable() by themselves as it will be removed from system boot up initialization.

### PR DESCRIPTION
**Pull request by eobalski**
This PR is for resolving point 1 from #13403
Closes #21419

### CODE REVIEW

**Review by carlescufi — COMMENTED** [review]
I think this is a reasonable way of solving this.
We should in theory deprecate `usb_enable()` since that is a public API, but in practice this is probably never been used before.

@jfischer-phytec-iot and @finikorg can you also state your opinion?

File: include/usb/usb_device.h
Comment (pabigot) [inline include/usb/usb_device.h]: Please remove `USB_DEVICE_NUM`.  Unless there's a reason why we need to iterate through enum values there's no need to name the upper bound, and doing so is confusing as there is no `NUM` USB device state.

File: subsys/usb/usb_device.c
Comment (pabigot) [inline subsys/usb/usb_device.c]: s/clalback/callback/g

File: subsys/usb/usb_device.c
Comment (pabigot) [inline subsys/usb/usb_device.c]: Was this intended to map `enum usb_dc_status_code` to `enum usb_state`?  If not, we can drop the commit that adds that enum.

**Review by pabigot — CHANGES_REQUESTED** [review]
As implemented the new USB device state enumeration is not used.  It should be removed, or the DC states mapped to device states.  This is the blocker, the rest is commentary.

I still don't like the fact that there's a single callback, which means every application that uses USB has to track the state and fan it out to modules that care about it.  I'd prefer to have something like:
```
struct usb_state_callback_node {
    sys_snode_t node;           /* internal use */
    usb_dc_status_callback cb;  /* initialize to user callback */
};

/** Link in a state-change callback.
 * @param node pointer to a persistent object initialized to contain a user callback.
 * Return: The current USB state, or a negative error code. */
int usb_state_add_callback(struct usb_state_callback_node *node);

/** Unlink a state-change callback. */
int usb_state_remove_callback(struct usb_state_callback_node *node);
```

Note that once there's an option to add a callback after USB has been initialized we need a way to ensure that the application tracks the state correctly: either by returning the state, or by having the callback invoked with an initial state when it's registered.

But I guess that's complex, and if nobody else thinks it's worthwhile I can live with doing the status fan-out in each application.


For the record the sequence I get on startup on NRF52840 is:
```
Status: 2 CONNECTED
Status: 5 SUSPEND
Status: 6 RESUME
Status: 1 RESET
Status: 1 RESET  // twice?
Status: 3 CONFIGURED
```

When the USB is unplugged I get:
```
Status: 5 SUSPEND
Status: 4 DISCONNECTED
```

If I plug in a power-only line I get:
```
Status: 2 CONNECTED
Status: 5 SUSPEND
```

which is a little odd.  It's not obvious how those map to device states (which should be extended to include a "DISCONNECTED" state, if that enum is going to be used).

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Comment by jfischer-phytec-iot** [comment]
@carlescufi @emob-nordic @finikorg After reviewing this and #20011 I think we should not introduce USB_DEVICE_AUTO_ENABLE and drop 7bc65fabccc57cb8d4235ac9a8d2a604b33ba856

File: include/usb/usb_device.h
Comment (jfischer-phytec-iot) [inline include/usb/usb_device.h]: usb_device_state?

File: subsys/usb/usb_device.c
Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: [usb_dc_status_code](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/usb/usb_dc.h#L41) should be reordered

**Comment by jfischer-phytec-iot** [comment]
> Note that once there's an option to add a callback after USB has been initialized we need a way to ensure that the application tracks the state correctly: either by returning the state, or by having the callback invoked with an initial state when it's registered.

This callback is used for notification only. The stack does not have to care about what application does with that. btw, I think a message queue or mailbox would be a better solution, but needs more work.

File: include/usb/usb_device.h
Comment (emob-nordic) [reply include/usb/usb_device.h]: Device can be resumed from suspend and end up in not only configured state. From Suspend You can either enter Default, Address or Configured State. I had a trouble describing Configured state in one sentence because as I refered to Table 9-1 from usb 2.0 spec (chapter 9)
_"Device is attached to the USB, powered, has been reset, has a unique address, is configured, and is not suspended. The host may now use the function provided by the device."_

Maybe it would be better something like:
_"USB device is configured and the host may use the function provided by the device"_ ?

File: include/usb/usb_device.h
Comment (emob-nordic) [reply include/usb/usb_device.h]: Yes, will update it.

File: subsys/usb/usb_device.c
Comment (emob-nordic) [reply subsys/usb/usb_device.c]: I was following the existing style it is present to this file. Same was done to [usb_register_status_callback](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/usb_device.c#L929). I wanted to be consistent but of course it could be simplified, I do not insist.

File: subsys/usb/usb_device.c
Comment (emob-nordic) [reply subsys/usb/usb_device.c]: Adding this enum I intended to provide to user the **state** enum type which will match states described in spec and nothing more. The callback is still using **statuses** as my intention was that the user is responsible for controlling the state.

Comment (emob-nordic) [reply subsys/usb/usb_device.c]: > [usb_dc_status_code](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/usb/usb_dc.h#L41) should be reordered

Do You mean the order of statuses should match those from [usb_state](https://github.com/zephyrproject-rtos/zephyr/pull/20375/files#diff-0a5e6a18914d2e2bf1bf9595e8ab483eR105) ?

**Comment by emob-nordic** [comment]
>As implemented the new USB device state enumeration is not used. It should be removed, or the DC states mapped to device states. This is the blocker, the rest is commentary.

The usb_device_state was prepared for users so they will not create their own enum for this. Nothing more.

I understand that ideally user_state_callback should use usb_device_state as an input parameter but:
- from what I checked it would only apply for nordic (some vendors forward only 2 statuses which is weird but it is done like this and its not my intention here to correct drivers).

Ideally We would have field device_state in usb_dev singleton and then based on **status** callback we could change the **state** internally in the device and then call  proper user callback to keep him informed about the **state**.

Just to clarify the view of this PR. What basically happens here is:
- throw away the hid_ops.status_cb and makes it global per device, not per class as this should never be class specific.

> which is a little odd. It's not obvious how those map to device states (which should be extended to include a "DISCONNECTED" state, if that enum is going to be used).

usb_state will not include DISCONNECTED state as its not consistent with the spec and for this I would not agree with You.

**Comment by pabigot** [comment]
> Just to clarify the view of this PR. What basically happens here is:
> 
> * throw away the hid_ops.status_cb and makes it global per device, not per class as this should never be class specific.

That's a good goal.

> The usb_device_state was prepared for users so they will not create their own enum for this. Nothing more.
> usb_state will not include DISCONNECTED state as its not consistent with the spec

I don't see the value in providing this enum if it's not going to be handled by the driver.  To this point AFAIK no classes have had any use for these states.  Even if they did, your point that some vendor implementations don't provide enough detail through DC states to distinguish them suggests they aren't generally useful.

Whatever receives the state changes will still have to interpret the DC states provided in the callback and translate them into whatever states are important.  For me that's "connected/working" and "disconnected/not working".  I do not care about the finer-grained states in `enum usb_state`.

I need a compelling reason why `enum usb_state` should remain, or fairly consistent agreement from the "real" USB maintainers that it should be here even if it's not used, before I withdraw the NAK.

**Comment by emob-nordic** [comment]
> I need a compelling reason why `enum usb_state` should remain, or fairly consistent agreement from the "real" USB maintainers that it should be here even if it's not used, before I withdraw the NAK.

I've discussed it internally and I agree in this matter. It's not common to all drivers (unfortunate) so shall not be there.

**Review by pabigot — APPROVED** [review]

File: samples/subsys/usb/hid-cdc/src/main.c
Comment (finikorg) [inline samples/subsys/usb/hid-cdc/src/main.c]: So to use this callback we shall call usb_enable(cb) always?

Comment (emob-nordic) [reply samples/subsys/usb/hid-cdc/src/main.c]: Yes. If callback is needed then USB_DEVICE_AUTO_ENABLE=n and call usb_enable(cb) in the app.

**Review by finikorg — APPROVED** [review]

File: include/usb/usb_device.h
Comment (jfischer-phytec-iot) [reply include/usb/usb_device.h]: No, usual it is Configured->Suspend->Configured. Not at all something like Configured->Suspend->Address.

File: subsys/usb/usb_device.c
Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: Please simplify :-)

File: subsys/usb/usb_device.c
Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: Yes, otherwise it will not work

Comment (emob-nordic) [reply subsys/usb/usb_device.c]: Right now I dont think there is a need to change anything in the  [usb_dc_status_code](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/usb/usb_dc.h#L41) as I agreed to remove usb_device_state (justification below)

Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: :confused: But was it not what @carlescufi and you wanted?

File: samples/subsys/usb/hid-cdc/src/main.c
Comment (jfischer-phytec-iot) [inline samples/subsys/usb/hid-cdc/src/main.c]: `if (usb_enable(status_cb))`?

File: samples/subsys/usb/hid-mouse/src/main.c
Comment (jfischer-phytec-iot) [inline samples/subsys/usb/hid-mouse/src/main.c]: `if (usb_enable(status_cb))`?

File: samples/subsys/usb/hid/src/main.c
Comment (jfischer-phytec-iot) [inline samples/subsys/usb/hid/src/main.c]: `if (usb_enable(status_cb))`?

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: samples/subsys/usb/hid-cdc/src/main.c
Comment (pabigot) [reply samples/subsys/usb/hid-cdc/src/main.c]: Personal opinion: yeah, it could be[*], but that would make it slightly more inconvenient to add diagnostics to display exactly why it failed when debugging or maintaining.

[*] well, actually this:

```
if (usb_enable(status_cb) != 0) {
```

since we're courting MISRA these days and treating `int` values as booleans is not acceptable.

File: subsys/usb/usb_device.c
Comment (carlescufi) [reply subsys/usb/usb_device.c]: What we wanted is to expose the Device States from the spec to the application. I disagree with @pabigot here, and I think we should expose them all. I see no need to just expose a subset.

Comment (pabigot) [reply subsys/usb/usb_device.c]: To clarify: my objection was to adding `enum usb_state` as a distinct type but not providing the translation from the existing DC states to them when passing them to the callback.  If you want to do the translation, that's fine, but from what I understand it would only work on Nordic because the other implementations can't provide the necessary detail.

Also, I do believe that "disconnected" is a state I need to be informed about (distinct from suspended), even if the spec doesn't consider it a named "device state".

File: include/usb/usb_device.h
Comment (emob-nordic) [reply include/usb/usb_device.h]: I meant depending in which state the device was before after entering Suspended it can either end up in x = (Powered, Default, Address, Configured). It's possible to have those transitions:
x -> Suspended -> x

File: samples/net/wpanusb/src/wpanusb.c
Comment (pfalcon) [inline samples/net/wpanusb/src/wpanusb.c]: What is sb_enable()? This may be an illustration why changes like this don't scale.

Comment (emob-nordic) [reply samples/net/wpanusb/src/wpanusb.c]: Sorry for that. I have to remove it by mistake when rebasing.

File: subsys/usb/usb_device.c
Comment (carlescufi) [inline subsys/usb/usb_device.c]: This should be done after the built-in classes have been informed

**Review by carlescufi — APPROVED** [review]
I see no other clean way to do this since we are already exposing the application to `usb_dc_status` with `usb_cfg_data`.

File: subsys/usb/usb_device.c
Comment (pabigot) [inline subsys/usb/usb_device.c]: The above check for `usb_dev.enabled` must be moved after the mutex is acquired, or two threads could pass that state, then sequentially both initialize the device.

**Review by pabigot — CHANGES_REQUESTED** [review]
The `usb_dev.enabled` check must be moved into the mutex region.

File: subsys/usb/usb_device.c
Comment (emob-nordic) [reply subsys/usb/usb_device.c]: Good point. I missed this case.

Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: Guess it was my fault too.

**Comment by emob-nordic** [comment]
@pabigot, @jfischer-phytec-iot  I've updated suggested changes. Please check it once again and approve if it's fine.

File: subsys/usb/usb_device.c
Comment (pabigot) [inline subsys/usb/usb_device.c]: No, this will produce deadlocks for anything that subsequently invokes `usb_enable()`.  The body has to be:
```
    ret = 0;
    goto out;
```
where the label `usb_enable_error` is renamed to `out`.

**Review by pabigot — DISMISSED** [review]
Need to ensure mutex is unlocked when leaving the mutex region.

File: subsys/usb/usb_device.c
Comment (emob-nordic) [reply subsys/usb/usb_device.c]: Done.

File: subsys/usb/usb_device.c
Comment (pabigot) [inline subsys/usb/usb_device.c]: I would like this to be renamed to `out` as it is the exit path for success as well.

File: subsys/usb/usb_device.c
Comment (pabigot) [inline subsys/usb/usb_device.c]: To be consistent with the documented behavior of `usb_enable()` this should be added, in case some of the intermediate functions returned a positive value:

    ret = 0;

File: subsys/usb/usb_device.c
Comment (pabigot) [inline subsys/usb/usb_device.c]: Here's one where I'm not responsible for this, so I won't block, but: This is unilaterally changing existing behavior of a subsystem, requiring code to change so the application continues to work.  There's no documentation of this change.  At a minimum it should be part of a release note on changed API.  Somebody should figure out how to make that work.

**Review by pabigot — DISMISSED** [review]
One blocking (ensuring `usb_enable()` returns zero as documented when it succeeds), one suggestion, and one note on the implicit, significant change in API that now requires the application to enable the device.

When I first started reviewing this PR I remember the intent was to have a Kconfig that controlled auto-enable.  I would be willing to approve this if that was restored and defaulted to auto-enable, or if there was a clear documented policy that allowed the change and explained how it would be communicated to users, but otherwise I'll stick to "won't reject".  I feel strongly that Zephyr must be much better about not breaking existing code even if the API is "unstable".

File: subsys/usb/usb_device.c
Comment (emob-nordic) [reply subsys/usb/usb_device.c]: The story behind this change starts from Nordic application team. There was a problem with missed status_callbacks in [hid_ops.status_cb](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/usb/class/usb_hid.h#L85) Two issues here:
1) status callback should never been associated with class.
2) usb_enable() calls usb_dc_attach() and by that call interrupts are enabled. Because it happens on SYS_INIT and user app is registering HID callbacks by itself (after system is woken up), it is possible to get the interrupt from usbd and in particular miss some statuses in the APP.

By this change we let the user to control usb_enable function and by that avoid those misses.

File: subsys/usb/usb_device.c
Comment (emob-nordic) [reply subsys/usb/usb_device.c]: To be honest I dont get it. if failure, then return success?

Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: API documented behavior for `usb_enable()` is  _"0 on success, negative errno code on fail."_, so it is undefined for (ret > 0) :-)

But as we are here, should we cleanup and change this` if (ret < 0) `to `if (ret)`?
- `static int usb_vbus_set(bool on);` returns 0 on success, negative errno code on fail
- `int usb_dc_attach(void); `returns 0 on success, negative errno code on fail
- `int usb_dc_ep_configure(const struct usb_dc_ep_cfg_data * const cfg);` returns 0 on success, negative errno code on fail
- `int usb_dc_ep_set_callback(const u8_t ep, const usb_dc_ep_callback cb);` returns 0 on success, negative errno code on fail
- `static int composite_setup_ep_cb(void); `returns 0 on success, negative errno code on fail
- `int usb_dc_ep_enable(const u8_t ep);` returns 0 on success, negative errno code on fail

Comment (pabigot) [reply subsys/usb/usb_device.c]: > API documented behavior for `usb_enable()` is _"0 on success, negative errno code on fail."_, so it is undefined for (ret > 0) :-)

Exactly.  Failing to describe the complete value range of input and output parameters is pervasive problem in Zephyr.

> But as we are here, should we cleanup and change this`if (ret < 0)`to `if (ret)`?

Changing to `if (ret != 0)` would be acceptable as it satisfies the return guarantee--as long as none of those functions return positive.  Making a maintainer search for the documented (and implemented) behavior of all those subordinate functions to confirm the API  is not reasonable IMO, hence the change request.

Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: @emob-nordic Do you agree with "Changing to if (ret != 0) would be acceptable" ?

Comment (emob-nordic) [reply subsys/usb/usb_device.c]: Yes, now I get the point. I will update it in a minute.

File: subsys/usb/usb_device.c
Comment (pabigot) [reply subsys/usb/usb_device.c]: By this change you **require** the user to control `usb_enable()`, causing all existing USB applications to break until they are modified to enable USB explicitly, just as you've done for the examples.  Or am I wrong about that?

I understand the reason for the change but a comment here in the PR doesn't help the maintainers of existing code figure out why their applications no longer work.

Comment (emob-nordic) [reply subsys/usb/usb_device.c]: @pabigot I will update documentation in commit in which I am changing from _auto enable_ to _enable by user_ as well as PR message. Would this be sufficient?

Comment (pabigot) [reply subsys/usb/usb_device.c]: The [documentation](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#deprecated) has a process.

I was going to say something about it assuming the API can be deprecated (rather than changed immediately) and focuses on stable APIs, but I looked at the [API Status](https://github.com/zephyrproject-rtos/zephyr/projects/15) board and USB device is marked as stable.

@carlescufi does this change warrant an exception?

Comment (carlescufi) [reply subsys/usb/usb_device.c]: @pabigot you raise a good point but here are the circumstances around this that I'd like to point out to warrant an exception in this particular case:

- The current behavior was inconsistent with other protocol stacks in Zephyr, and in particular with Bluetooth, that requires a `bt_enable()` call from the sample that controls the moment at which we power on the hardware.
- The `usb_enable()` function was almost useless in the form that existed until this PR, and we are now giving it a purpose.
- I agree we need to document this behavior in two places: The doxygen for `usb_enable()` itself and the release notes (2.2 release notes at this point).
- We will merge this change right in the beginning of the 2.2 release cycle, which give ample time to users that follow master to adapt.
- Deprecating the current behavior would not benefit users in any meaningful way, because of the current inconsistent and incorrect behavior of this `usb_enable()` API.
- For users of standard provided classes that expose an API, eg. HID there was a race condition with the current master behavior of auto-enabling the USB stack: the USB stack was enabled before the user had a chance to register a HID device, meaning that events could be lost in between.

**Review by jfischer-no — DISMISSED** [review]

**Review by pabigot — COMMENTED** [review]
I like (and need) the new capability, but can't approve the API breakage being merged with no attempt to follow the standard deprecation process or to extend the process to cover cases where deprecation is simply infeasible.

**Comment by carlescufi** [comment]
As per the API meeting, @emob-nordic please bring back the Kconfig option to conditionally disable the auto-powerup of the USB stack into this PR.

**Comment by jfischer-phytec-iot** [comment]
> As per the API meeting, @emob-nordic please bring back the Kconfig option in #20011 to conditionally disable the auto-powerup of the USB stack into this PR.

Disagree, that is not the problem, the cb argument breaks the API.

**Review by jfischer-no — CHANGES_REQUESTED** [review]
We forgot to rework netusb, please take a look/cherry-pick https://github.com/jfischer-phytec-iot/zephyr/tree/20375

**Comment by emob-nordic** [comment]
> > As per the API meeting, @emob-nordic please bring back the Kconfig option in #20011 to conditionally disable the auto-powerup of the USB stack into this PR.
> 
> Disagree, that is not the problem, the cb argument breaks the API.
> But there is more to discuss, this PR breaks also netusb support.

I will leave this discussion for USB sync meeting today.

**Comment by carlescufi** [comment]
@emob-nordic  please follow [the new procedure](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes) and let's not introduce the Kconfig option, taking the opportunity to clean this up since we are introducing incompatible changes anyway.

File: doc/releases/release-notes-2.2.rst
Comment (dbkinder) [inline doc/releases/release-notes-2.2.rst]: ``function`` misspelled

**Review by dbkinder — DISMISSED** [review]
Seems like there should be some usage documentation changes for this (other than in the doxygen comments and release notes)...

**Comment by emob-nordic** [comment]
> Seems like there should be some usage documentation changes for this (other than in the doxygen comments and release notes)...

What would You suggest? I was following the [process](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes) and hoped this is sufficient.

File: doc/releases/release-notes-2.2.rst
Comment (carlescufi) [inline doc/releases/release-notes-2.2.rst]: I will give a shot at another description here @dbkinder and @emob-nordic:
------
* The `usb_enable()` function, which was previously invoked automatically by the USB stack, now needs to be explicitly called by the application in order to enable the USB subsystem.
* The `usb_enable()` function now takes a parameter,  `usb_dc_status_callback` which can be set by the application to a callback to receive status events from the USB stack. The parameter can also be set to `NULL` if no callback is required.

----

**Comment by carlescufi** [comment]
> Seems like there should be some usage documentation changes for this (other than in the doxygen comments and release notes)...

I proposed a rewritten section in the release notes. Beyond this and the doxygen changes, there is currently no documentation which explains to the users the USB subsystem usage, so I do not think any further changes are necessary until we add more documentation @dbkinder.
The change follows standard procedure as pointed out by @emob-nordic

File: doc/releases/release-notes-2.2.rst
Comment (StefJar) [reply doc/releases/release-notes-2.2.rst]: lovely - I like it. IMHO this is how it should be

**Review by jfischer-no — APPROVED** [review]

File: samples/bluetooth/hci_usb/src/main.c
Comment (joerchan) [inline samples/bluetooth/hci_usb/src/main.c]: blank space after `}`

**Review by joerchan — APPROVED** [review]

**Pull request merged**
